### ISSUE TITLE
stories.comments_count 1 when story has 2 comments, submitted simultaneously.

### ISSUE THREAD

**Issue by alanpost** [lobsters/lobsters#813]
As of writing this ticket, the following story has 2 comments whereas the summary line reports 1 comment:

https://lobste.rs/s/hmzsdm/make_resilient_go_net_http_servers_using
Make resilient Go net/http servers using timeouts, deadlines and context cancellation
1 comment

With two comments, in truth one a duplicate of the other, submitted simultaneously and by accident:

https://lobste.rs/s/hmzsdm/make_resilient_go_net_http_servers_using#c_qynkp9
https://lobste.rs/s/hmzsdm/make_resilient_go_net_http_servers_using#c_zn8fmh

The database confirms this condition:

    MariaDB [lobsters]> select short_id, comments_count from stories where short_id="hmzsdm";
    +----------+----------------+
    | short_id | comments_count |
    +----------+----------------+
    | hmzsdm   |              1 |
    +----------+----------------+

    MariaDB [lobsters]> select short_id, created_at, "hmzsdm" from comments where story_id=(select id from stories where short_id="hmzsdm");               
    +----------+---------------------+--------+
    | short_id | created_at          | hmzsdm |
    +----------+---------------------+--------+
    | qynkp9   | 2020-01-25 12:53:22 | hmzsdm |
    | zn8fmh   | 2020-01-25 12:53:22 | hmzsdm |
    +----------+---------------------+--------+

Looking at the log data, I speculate what happened is that the first comment POST, while running update_comments_count, ran after the second, "winning" the race to the database.  You can see each request being processed in parallel with the original request (e581ad7c-24e2-4282-9784-71dc71e98618) ending last:

    I, [2020-01-25T12:53:22.153829 #21736]  INFO -- : [e581ad7c-24e2-4282-9784-71dc71e98618] Started POST "/comments" for 1.2.3.4 at 2020-01-25 12:53:22 +0000
    I, [2020-01-25T12:53:22.153910 #21688]  INFO -- : [b868c86b-34c3-407d-9f26-87a9e48cce55] Started POST "/comments" for 1.2.3.4 at 2020-01-25 12:53:22 +0000
    I, [2020-01-25T12:53:22.155466 #21736]  INFO -- : [e581ad7c-24e2-4282-9784-71dc71e98618] Processing by CommentsController#create as */*
    I, [2020-01-25T12:53:22.155473 #21688]  INFO -- : [b868c86b-34c3-407d-9f26-87a9e48cce55] Processing by CommentsController#create as */*
    I, [2020-01-25T12:53:22.155621 #21736]  INFO -- : [e581ad7c-24e2-4282-9784-71dc71e98618]   Parameters: {"story_id"=>"hmzsdm", ...}
    I, [2020-01-25T12:53:22.155717 #21688]  INFO -- : [b868c86b-34c3-407d-9f26-87a9e48cce55]   Parameters: {"story_id"=>"hmzsdm", ...}
    I, [2020-01-25T12:53:22.158841 #21688]  INFO -- : [b868c86b-34c3-407d-9f26-87a9e48cce55]   Request 1.2.3.4 POST /comments user: 10226 arp242
    I, [2020-01-25T12:53:22.165128 #21736]  INFO -- : [e581ad7c-24e2-4282-9784-71dc71e98618]   Request 1.2.3.4 POST /comments user: 10226 arp242
    I, [2020-01-25T12:53:22.214030 #21688]  INFO -- : [b868c86b-34c3-407d-9f26-87a9e48cce55] Completed 200 OK in 58ms (Views: 5.8ms | ActiveRecord: 19.8ms)
    I, [2020-01-25T12:53:22.242983 #21736]  INFO -- : [e581ad7c-24e2-4282-9784-71dc71e98618] Completed 200 OK in 87ms (Views: 6.1ms | ActiveRecord: 42.7ms)

**Comment by alanpost** [comment]
Thank you to user enpo in #lobsters on freenode for noticing and reporting this discrepancy.

**Comment by adrianke** [comment]
I found another duplicate comment today.

I propose the following solution:
 1. On page load, the application generates a unique string (idempotency token) and embeds it into the page.
 2. When the user tries to submit the value is added into the database in its own column.
 3. The column has a UNIQUE CONSTRAINT.

Duplicate comments would thus be discarded by the database.

**Comment by pushcx** [comment]
I'm making a pass through old issues and I'm closing this out. We've been replacing our custom stats because of bugs like this, and will entirely remove the Keystore system in #1585.
